semgrep-core/src/matching/Generic_vs_generic.ml

```diff
@@ -494,7 +506,29 @@ let rec m_name a b =
       (* this will handle metavariables in Id *)
       m_ident_and_id_info (a1, a2) (b1, b2)
   | G.Id ((str, tok), _info), G.IdQualified _ when MV.is_metavar_name str ->
-      envf (str, tok) (MV.N b)
+      (* If `is_resolved` is true, then we got the target `IdQualified` from a
+         `ResolvedName`.
+         Such an identifier has a nonsensical range, because it may have been
+         constructed from identifiers from arbitrary areas, such as imports or
+         classes.
+
+         If the resolved name is `A.foo`, in
+         the following code:
+         ```
+         class A:
+           def foo():
+             pass
+         ```
+         then the pattern metavariable will match the range of the text
+         ```
+         A:
+         def foo
+         ```
+         because it is matching everything between `A` and `foo`.
+
+         So we should not permit this match.
+      *)
+      if is_resolved then fail () else envf (str, tok) (MV.N b)
```

Comment:
Does it make sense to just check whether the tokens in `b` are coherent? We could tag `IdQualified`s as "generated" (vs coming from the sources) and that would make it trivial to check.

Reply:
My next PR (on top of this one) will make _all_ such tokens coherent, so no point to adding a check